tiledb/sm/query/readers/sparse_global_order_reader.cc

```diff
@@ -440,51 +941,258 @@ SparseGlobalOrderReader<BitmapType>::create_result_tiles(
+                    ", num fragments to process " +
+                    std::to_string(num_fragments_to_process)));
+              }
+              return Status::Ok();
+            }
+          }
+
+          tmp_read_state_.set_all_tiles_loaded(f);
+
+          return Status::Ok();
+        }));
   }
+}
 
-  logger_->debug("Done adding result tiles, num result tiles {0}", num_rt);
+template <class BitmapType>
+void SparseGlobalOrderReader<BitmapType>::create_result_tiles_using_preprocess(
+    std::vector<ResultTilesList>& result_tiles,
+    std::optional<PreprocessTileMergeFuture>& merge_future) {
+  // For easy reference.
+  const auto num_dims = array_schema_.dim_num();
+
+  if (preprocess_tile_order_.cursor_ > 0 &&
+      std::all_of(result_tiles.begin(), result_tiles.end(), [](const auto& r) {
+        return r.empty();
+      })) {
+    // When a result tile is created for a ResultTileId, the preprocess merge
+    // cursor advances. Then the result tile tracks how much data has been
+    // emitted from that tile.
+    //
+    // When running natively, if an iteration does not exhaust that result tile
+    // remains in the result tile list and its cursor remains in memory for
+    // use in the next iteration.
+    //
+    // When running against the REST server, if an iteration does not exhaust a
+    // result tile, then its cursor must be serialized/deserialized.  That state
+    // lives in `read_state_.frag_idx()`.
+    //
+    // Hence if we have no result tiles, and a preprocess cursor, then we must
+    // check if we have any un-exhausted result tiles.
+    const auto& rtcursors = read_state_.frag_idx();
+
+    if (merge_future.has_value()) {
+      const bool wait_ok =
+          merge_future->wait_for(preprocess_tile_order_.cursor_ - 1);
+      if (!wait_ok) {
+        throw SparseGlobalOrderReaderInternalError(
+            "Waiting for preprocess tile merge results failed");
+      }
+    }
 
-  if (done_adding_result_tiles) {
-    logger_->debug("All result tiles loaded");
+    for (const auto& f : subarray_.relevant_fragments()) {
+      // identify the tiles in reverse global order
+      std::vector<size_t> rts;
+      for (size_t rti = preprocess_tile_order_.cursor_; rti > 0; --rti) {
+        const auto& rt = preprocess_tile_order_.tiles_[rti - 1];
+        if (rt.fragment_idx_ != f) {
+          continue;
+        } else if (rtcursors[f].tile_idx_ > rt.tile_idx_) {
+          // all of the tiles for this fragment up to this are done
+          break;
+        }
+
+        rts.push_back(rti - 1);
+      }
+
+      // and then re-create them
+      for (auto rti = rts.rbegin(); rti != rts.rend(); ++rti) {
+        const auto& rt = preprocess_tile_order_.tiles_[*rti];
+
+        // this is a tile which qualified for the subarray and was
+        // a created result tile, we must continue processing it
+        bool budget_exceeded;
+        while ((budget_exceeded = add_result_tile(
```

Comment:
This line caused a warning-turned-error in the backport PR on VS2019 (which I suppressed). The `add_result_tile` function is pre-existing, but I must say that returning a boolean indicating whether the budget was exceeded is very confusing.